hybrid/per_func_comp_main.c

```diff
@@ -0,0 +1,257 @@
+
+    void *__capability sealcap;
+    sealcap = cheri_maketype(global_sealcap, tysz);
+    __builtin_cheri_cap_type_copy(global_sealcap, sealcap);
+    return sealcap;
+}
+
+/*
+//
+// TODO: test function attributes (ccall, ccallee), with args
+//
+__attribute__((cheri_ccall))
+__attribute__((cheri_method_suffix("_cap")))
+__attribute__((cheri_method_class(main_obj)))
+int sandbox_1_func(int * __capability intcap);
+
+int sandbox_1_func(int * __capability intcap)
+{
+    struct cheri_object sb1_obj;
+
+    sb1_obj.co_codecap = cheriobj_1_codecap;
+    sb1_obj.co_datacap = cheriobj_1_datacap;
+
+    (void)libcheri_invoke(sb1_obj, 0,
+            0, 0, 0, 0, 0, 0, 0, 0,
+            NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
+
+    return *intcap + 1;
+}
+
+__attribute__((cheri_ccall))
+__attribute__((cheri_method_suffix("_cap")))
+__attribute__((cheri_method_class(sandbox2_obj)))
+int sandbox_2_func(int * __capability intcap);
+
+int sandbox_2_func(int * __capability intcap)
+{
+    struct cheri_object sb2_obj;
+
+    sb2_obj.co_codecap = cheriobj_2_codecap;
+    sb2_obj.co_datacap = cheriobj_2_datacap;
+
+    (void)libcheri_invoke(sb2_obj, 0,
+            0, 0, 0, 0, 0, 0, 0, 0,
+            NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
+
+    return *intcap + 1;
+}
+///
+/// up_to_here
+///
+*/
+
+void sandbox_1_func()
+{
+    printf("--> end_of_sandbox_1_func\n");
+}
+
+void sandbox_2_func()
+{
+    printf("--> end_of_sandbox_2_func\n");
+}
+
+int main() // Remove unused args
+{
+    printf("----------------------------------------\n");
+    printf("CHERI_OTYPE_USER_MIN: %04x\n", CHERI_OTYPE_USER_MIN);
+    printf("CHERI_OTYPE_USER_MAX: %04x\n", CHERI_OTYPE_USER_MAX);
+    printf("----------------------------------------\n\n");
+
+    //=- Init a global_sealing capability -=//
+
+    size_t global_sealcap_sz = sizeof(global_sealcap); // sizeof(global_sealcap);
+    if (sysctlbyname("security.cheri.sealcap", &global_sealcap, &global_sealcap_sz, NULL, 0) < 0)
+    {
+        printf("Fatal error. Cannot get `security.cheri.sealcap`.");
+        exit(1);
+    }
+    //=- Create cheri_object_1 sealcap -=//
+
+    //- Get a next type i.e. create a next sealing capability -//
+    size_t sealcap_size_1 = 12; // --> Random size is given to test how types are assigned
+    void *__capability sealcap_1 = alloc_type(sealcap_size_1);
+
+    //- NOTE: Since this example is just to check otypes,
+    //- we set arbitrary address ranges of code and data capability
+
+    //- Create cheri_object_1 *code* cap -//
+
+    void *sandbox_code_1_ubound = (void *) (((size_t) sandbox_1_func) + SANDBOX_CODE_SZ);
+    cheriobj_1_codecap = cheri_seal(
+        codecap_create((void (*)(void)) & sandbox_1_func, sandbox_code_1_ubound), sealcap_1);
+
+    //- Create cheri_object_1 *data* cap -//
+
+    //- Two cheri_objects' data addr ranges are set overlapped. Ignore the numbers for this example.
+    //-//
+    void *apprx_data_UB = __builtin_frame_address(0);
```

Comment:
I don't get what this `__builtin_frame_address` is meant to do; is it somehow related to comp data spaces? Also it seems `cheriobj_2_data_base` and `cheriobj_data_1_base` are the same?

Related, I assume these two variables refer to the same thing across the two compartments? Could we have consistent naming throughout?

Reply:
I think that's future work for cheri-based user defined compartmentalization - setting ranges of addresses. Ignore for now, as I commected there.